test/SILOptimizer/protocol_devirtulize.sil

```diff
@@ -0,0 +1,54 @@
+// RUN: %target-sil-opt -wmo -assume-parsing-unqualified-ownership-sil -enable-sil-verify-all %s -protocol-devirtualizer -inline -sil-combine -generic-specializer -devirtualizer -late-inline -dead-arg-signature-opt -dce -sil-deadfuncelim | %FileCheck %s
+//
+import Builtin
+import Swift
+
+internal protocol SomeProtocol : AnyObject {
+  func increment() -> Int
+}
+
+internal class SomeClass : SomeProtocol {
+  init()
+  func increment() -> Int
+}
+
+// CHECK-LABEL: something
+// CHECK: bb0
+// CHECK: alloc_ref
+// CHECK-NOT: init_existential_ref
+// CHECK: function_ref @$S25something_to_devirtualizeTf5n_n4main9SomeClassC_Tg5
+sil @something : $@convention(thin) () -> Int {
+bb0:
+  %0 = alloc_ref $SomeClass
+  %1 = init_existential_ref %0 : $SomeClass : $SomeClass, $SomeProtocol
+  %2 = function_ref @something_to_devirtualize : $@convention(thin) (@guaranteed SomeProtocol) -> Int
+  %3 = apply %2(%1) : $@convention(thin) (@guaranteed SomeProtocol) -> Int
+  return %3 : $Int
+} 
+
+sil private [transparent] [thunk] @increment : $@convention(witness_method: SomeProtocol) (@guaranteed SomeClass) -> Int {
+bb0(%0 : $SomeClass):
+  %1 = integer_literal $Builtin.Int64, 10 
+  %2 = struct $Int (%1 : $Builtin.Int64) 
+  return %2 : $Int 
+} 
+
+// CHECK-LABEL: sil shared [noinline] @$S25something_to_devirtualizeTf5n_n4main9SomeClassC_Tg5 : $@convention(thin) (@guaranteed SomeClass) -> Int {
```

Comment:
I think instead of replacing the protocol parameter with the sole conforming class, it should remain a generic parameter. The generic specializer can specialize it to the class later.

As a bonus you should be able to get rid of the "sole conforming class" analysis. It makes the optimization only apply in very narrow cases, when it could be more general.

Reply:
related ticket, from the the ABI dashboard: https://bugs.swift.org/browse/SR-4331

Reply:
Just a heads up: the generic method and thunk creation now works for both protocol compositions and class-constrained compositions. I was able to eliminate the analysis altogether. I am working on the peephole optimizer right now. I will see if I can combine mine with the existing  self peephole optimizer.